File: qtfuzzylite/examples/shower.cpp
/* 
 * File:   shower.cpp
 * Author: jcrada
 *
 * Created on 12 January 2013, 11:26 PM
 */

#include <fl/Headers.h>

int main(int argc, char* argv[]) {
    fl::Engine* engine = new fl::Engine;
    engine->setName("shower");
    engine->addHedge(new fl::Any);
    engine->addHedge(new fl::Extremely);
    engine->addHedge(new fl::Not);
    engine->addHedge(new fl::Seldom);
    engine->addHedge(new fl::Somewhat);
    engine->addHedge(new fl::Very);

    fl::InputVariable* inputVariable1 = new fl::InputVariable;
    inputVariable1->setName("temp");
    inputVariable1->setRange(-20.000, 20.000);

    inputVariable1->addTerm(new fl::Trapezoid("cold", -30.000, -30.000, -15.000, 0.000));
    inputVariable1->addTerm(new fl::Triangle("good", -10.000, 0.000, 10.000));
    inputVariable1->addTerm(new fl::Trapezoid("hot", 0.000, 15.000, 30.000, 30.000));
    engine->addInputVariable(inputVariable1);

    fl::InputVariable* inputVariable2 = new fl::InputVariable;
    inputVariable2->setName("flow");
    inputVariable2->setRange(-1.000, 1.000);

    inputVariable2->addTerm(new fl::Trapezoid("soft", -3.000, -3.000, -0.800, 0.000));
    inputVariable2->addTerm(new fl::Triangle("good", -0.400, 0.000, 0.400));
    inputVariable2->addTerm(new fl::Trapezoid("hard", 0.000, 0.800, 3.000, 3.000));
    engine->addInputVariable(inputVariable2);

    fl::OutputVariable* outputVariable1 = new fl::OutputVariable;
    outputVariable1->setName("cold");
    outputVariable1->setRange(-1.000, 1.000);
    outputVariable1->setDefaultValue(std::numeric_limits<fl::scalar>::quiet_NaN());
    outputVariable1->setLockDefuzzifiedValue(true);
    outputVariable1->setDefuzzifier(new fl::Centroid(500));
    outputVariable1->output()->setAccumulation(new fl::Maximum);

    outputVariable1->addTerm(new fl::Triangle("closeFast", -1.000, -0.600, -0.300));
    outputVariable1->addTerm(new fl::Triangle("closeSlow", -0.600, -0.300, 0.000));
    outputVariable1->addTerm(new fl::Triangle("steady", -0.300, 0.000, 0.300));
    outputVariable1->addTerm(new fl::Triangle("openSlow", 0.000, 0.300, 0.600));
    outputVariable1->addTerm(new fl::Triangle("openFast", 0.300, 0.600, 1.000));
    engine->addOutputVariable(outputVariable1);

    fl::OutputVariable* outputVariable2 = new fl::OutputVariable;
    outputVariable2->setName("hot");
    outputVariable2->setRange(-1.000, 1.000);
    outputVariable2->setDefaultValue(std::numeric_limits<fl::scalar>::quiet_NaN());
    outputVariable2->setLockDefuzzifiedValue(true);
    outputVariable2->setDefuzzifier(new fl::Centroid(500));
    outputVariable2->output()->setAccumulation(new fl::Maximum);

    outputVariable2->addTerm(new fl::Triangle("closeFast", -1.000, -0.600, -0.300));
    outputVariable2->addTerm(new fl::Triangle("closeSlow", -0.600, -0.300, 0.000));
    outputVariable2->addTerm(new fl::Triangle("steady", -0.300, 0.000, 0.300));
    outputVariable2->addTerm(new fl::Triangle("openSlow", 0.000, 0.300, 0.600));
    outputVariable2->addTerm(new fl::Triangle("openFast", 0.300, 0.600, 1.000));
    engine->addOutputVariable(outputVariable2);

    fl::RuleBlock* ruleblock1 = new fl::RuleBlock;
    ruleblock1->setName("");
    ruleblock1->setTnorm(new fl::Minimum);
    ruleblock1->setSnorm(new fl::Maximum);
    ruleblock1->setActivation(new fl::Minimum);

    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is cold and flow is soft then cold is openSlow and hot is openFast", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is cold and flow is good then cold is closeSlow and hot is openSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is cold and flow is hard then cold is closeFast and hot is closeSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is good and flow is soft then cold is openSlow and hot is openSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is good and flow is good then cold is steady and hot is steady", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is good and flow is hard then cold is closeSlow and hot is closeSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is hot and flow is soft then cold is openFast and hot is openSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is hot and flow is good then cold is openSlow and hot is closeSlow", engine));
    ruleblock1->addRule(fl::MamdaniRule::parse(
            " if temp is hot and flow is hard then cold is closeSlow and hot is closeFast", engine));
    engine->addRuleBlock(ruleblock1);

}

File: examples/mamdani/matlab/shower.java
import com.fuzzylite.*;
import com.fuzzylite.defuzzifier.*;
import com.fuzzylite.factory.*;
import com.fuzzylite.hedge.*;
import com.fuzzylite.imex.*;
import com.fuzzylite.norm.*;
import com.fuzzylite.norm.s.*;
import com.fuzzylite.norm.t.*;
import com.fuzzylite.rule.*;
import com.fuzzylite.term.*;
import com.fuzzylite.variable.*;

public class shower{
public static void main(String[] args){
Engine engine = new Engine();
engine.setName("shower");

InputVariable inputVariable1 = new InputVariable();
inputVariable1.setEnabled(true);
inputVariable1.setName("temp");
inputVariable1.setRange(-20.000, 20.000);
inputVariable1.addTerm(new Trapezoid("cold", -30.000, -30.000, -15.000, 0.000));
inputVariable1.addTerm(new Triangle("good", -10.000, 0.000, 10.000));
inputVariable1.addTerm(new Trapezoid("hot", 0.000, 15.000, 30.000, 30.000));
engine.addInputVariable(inputVariable1);

InputVariable inputVariable2 = new InputVariable();
inputVariable2.setEnabled(true);
inputVariable2.setName("flow");
inputVariable2.setRange(-1.000, 1.000);
inputVariable2.addTerm(new Trapezoid("soft", -3.000, -3.000, -0.800, 0.000));
inputVariable2.addTerm(new Triangle("good", -0.400, 0.000, 0.400));
inputVariable2.addTerm(new Trapezoid("hard", 0.000, 0.800, 3.000, 3.000));
engine.addInputVariable(inputVariable2);

OutputVariable outputVariable1 = new OutputVariable();
outputVariable1.setEnabled(true);
outputVariable1.setName("cold");
outputVariable1.setRange(-1.000, 1.000);
outputVariable1.fuzzyOutput().setAccumulation(new Maximum());
outputVariable1.setDefuzzifier(new Centroid(200));
outputVariable1.setDefaultValue(Double.NaN);
outputVariable1.setLockValidOutput(false);
outputVariable1.setLockOutputRange(false);
outputVariable1.addTerm(new Triangle("closeFast", -1.000, -0.600, -0.300));
outputVariable1.addTerm(new Triangle("closeSlow", -0.600, -0.300, 0.000));
outputVariable1.addTerm(new Triangle("steady", -0.300, 0.000, 0.300));
outputVariable1.addTerm(new Triangle("openSlow", 0.000, 0.300, 0.600));
outputVariable1.addTerm(new Triangle("openFast", 0.300, 0.600, 1.000));
engine.addOutputVariable(outputVariable1);

OutputVariable outputVariable2 = new OutputVariable();
outputVariable2.setEnabled(true);
outputVariable2.setName("hot");
outputVariable2.setRange(-1.000, 1.000);
outputVariable2.fuzzyOutput().setAccumulation(new Maximum());
outputVariable2.setDefuzzifier(new Centroid(200));
outputVariable2.setDefaultValue(Double.NaN);
outputVariable2.setLockValidOutput(false);
outputVariable2.setLockOutputRange(false);
outputVariable2.addTerm(new Triangle("closeFast", -1.000, -0.600, -0.300));
outputVariable2.addTerm(new Triangle("closeSlow", -0.600, -0.300, 0.000));
outputVariable2.addTerm(new Triangle("steady", -0.300, 0.000, 0.300));
outputVariable2.addTerm(new Triangle("openSlow", 0.000, 0.300, 0.600));
outputVariable2.addTerm(new Triangle("openFast", 0.300, 0.600, 1.000));
engine.addOutputVariable(outputVariable2);

RuleBlock ruleBlock = new RuleBlock();
ruleBlock.setEnabled(true);
ruleBlock.setName("");
ruleBlock.setConjunction(new Minimum());
ruleBlock.setDisjunction(new Maximum());
ruleBlock.setActivation(new Minimum());
ruleBlock.addRule(Rule.parse("if temp is cold and flow is soft then cold is openSlow and hot is openFast", engine));
ruleBlock.addRule(Rule.parse("if temp is cold and flow is good then cold is closeSlow and hot is openSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is cold and flow is hard then cold is closeFast and hot is closeSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is good and flow is soft then cold is openSlow and hot is openSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is good and flow is good then cold is steady and hot is steady", engine));
ruleBlock.addRule(Rule.parse("if temp is good and flow is hard then cold is closeSlow and hot is closeSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is hot and flow is soft then cold is openFast and hot is openSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is hot and flow is good then cold is openSlow and hot is closeSlow", engine));
ruleBlock.addRule(Rule.parse("if temp is hot and flow is hard then cold is closeSlow and hot is closeFast", engine));
engine.addRuleBlock(ruleBlock);


}
}
